File: src_java/Upperbound.java
import java.util.List;

import datastructures.CLCSInstance;
import datastructures.Node;

public class Upperbound {

    public static int ubMinOccurrence(int sigmaLength, int numInputs, List<Integer> positionVector, int[] inputLengths,
            int[][][] occurrenceTables) {
        int ub = 0;

        boolean[] isOvershoot = getIsOvershoot(numInputs, positionVector, inputLengths);

        for (int label = 0; label < sigmaLength; label++) {
            int labelMin = Integer.MAX_VALUE;

            for (int i = 0; i < numInputs; i++) {
                int inputMin = isOvershoot[i] ? 0 : occurrenceTables[i][label][positionVector.get(i) - 1];
                labelMin = Math.min(labelMin, inputMin);
            }

            ub += labelMin;
        }

        return ub;
    }

    private static boolean[] getIsOvershoot(int numInputs, List<Integer> positionVector, int[] inputLengths) {
        boolean[] isOvershoot = new boolean[numInputs];

        for (int i = 0; i < numInputs; i++) {
            isOvershoot[i] = positionVector.get(i) > inputLengths[i];
        }
        return isOvershoot;
    }

    public static int ubMScore(int minLength, int numInputs, List<Integer> positionVector, int[][][] mScores) {
        int ub = minLength;
        for (int i = 0; i < numInputs - 1; i++) {
            ub = Math.min(ub, mScores[i][positionVector.get(i) - 1][positionVector.get(i + 1) - 1]);
        }
        return ub;
    }

    public static int ubBoth(CLCSInstance inst, Node v) {
        int ubMinOcc = ubMinOccurrence(inst.getSigmaLength(), inst.getNumInputs(), v.getPv(), inst.getInputLengths(),
                inst.getOccurrenceTables());

        int ubMscore = ubMScore(inst.getMinLength(), inst.getNumInputs(), v.getPv(), inst.getmScores());

        int ub = Math.min(ubMinOcc, ubMscore);
        return v.getL_v() + ub;
    }
}

File: src_py/upperbound.py
import numpy as np

from datastructure import Node, CLCS

INT_MAX = np.iinfo(np.int32).max

def ub_is_overshoot(num_inputs: int, pos_vec: list[int], input_lens: list[int]) -> list[bool]:
    return [pos_vec[i] > input_lens[i] for i in range(num_inputs)]

def ub_min_occurrence(sigma_len: int, num_inputs: int, pos_vec: list[int], input_lens: list[int], occurrence_tables: np.ndarray) -> int:
    is_overshoot = ub_is_overshoot(num_inputs, pos_vec, input_lens)
    ub = 0
    for label in range(sigma_len):
        label_min = INT_MAX
        
        for i in range(num_inputs):
            input_min = 0 if is_overshoot[i] else occurrence_tables[i][label, pos_vec[i] - 1]
            label_min = min(label_min, input_min)
        ub += label_min
    
    return ub

def ub_mscore(min_len: int, num_inputs: int, pos_vec: list[int], mscores: np.ndarray) -> int:
    ub = min_len
    
    for i in range(num_inputs-1):
        ub = min(ub, mscores[i][pos_vec[i] - 1, pos_vec[i+1] - 1])
    
    return ub

def ub_both(inst: CLCS, v: Node) -> int:
    ub_minc = ub_min_occurrence(inst.sigma_len, inst.num_inputs, v.pv, inst.input_lens, inst.occurrence_tables)
    ub_mscr = ub_mscore(inst.min_len, inst.num_inputs, v.pv, inst.mscores)
    
    ub = min(ub_minc, ub_mscr)
    
    return v.l_v + ub